shared-bindings/eveL/__init__.c

```diff
@@ -0,0 +1,237 @@
+
+//| :mod:`eveL` --- low-level BridgeTek EVE bindings
+//| ================================================
+//|
+//| .. module:: eveL
+//|   :synopsis: low-level BridgeTek EVE bindings
+//|   :platform: SAMD21/SAMD51
+//|
+//| The `eveL` module provides a class EVEL which
+//| contains methods for constructing EVE command 
+//| buffers and appending basic graphics commands.
+//|
+
+typedef struct _mp_obj_EVEL_t {
+    mp_obj_base_t base;
+    mp_obj_t writer;
+    mp_obj_t dest[3];
+
+    size_t n;
+    uint8_t buf[512];
+} mp_obj_EVEL_t;
+
+STATIC void _write(mp_obj_EVEL_t *EVEL, mp_obj_t b) {
+    EVEL->dest[2] = b;
+    mp_call_method_n_kw(1, 0, EVEL->dest);
+}
+
+STATIC mp_obj_t _register(mp_obj_t self, mp_obj_t o) {
+    mp_obj_EVEL_t *EVEL = self;
+    EVEL->n = 0;
+    mp_printf(&mp_plat_print, "register %p %d\n", EVEL, EVEL->n);
+    mp_load_method(o, MP_QSTR_write, EVEL->dest);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(register_obj, _register);
+
+STATIC mp_obj_t _flush(mp_obj_t self) {
+    mp_obj_EVEL_t *EVEL = self;
+    // mp_printf(&mp_plat_print, "flush %p %d\n", EVEL, EVEL->n);
+    if (EVEL->n != 0) {
+      _write(EVEL, mp_obj_new_bytearray_by_ref(EVEL->n, EVEL->buf));
+      EVEL->n = 0;
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(flush_obj, _flush);
+
+STATIC void *append(mp_obj_EVEL_t *EVEL, size_t m) {
+  if ((EVEL->n + m) > sizeof(EVEL->buf))
+    _flush((mp_obj_t)EVEL);
+  uint8_t *r = EVEL->buf + EVEL->n;
+  EVEL->n += m;
+  return (void*)r;
+}
+
+STATIC mp_obj_t _cc(mp_obj_t self, mp_obj_t b) {
+    mp_obj_EVEL_t *EVEL = self;
+    mp_buffer_info_t buffer_info;
+    mp_get_buffer_raise(b, &buffer_info, MP_BUFFER_READ);
+    // mp_printf(&mp_plat_print, "flush %p %d %p\n", EVEL, buffer_info.len, EVEL->writer);
+    if (buffer_info.len <= sizeof(EVEL->buf)) {
+      uint8_t *p = (uint8_t*)append(EVEL, buffer_info.len);
+      // memcpy(p, buffer_info.buf, buffer_info.len);
+      uint8_t *s = buffer_info.buf;
+      for (size_t i = 0; i < buffer_info.len; i++)
+          *p++ = *s++;
+    } else {
+      _flush(self);
+      _write(EVEL, b);
+    }
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(cc_obj, _cc);
+
+#define C4(self, u) (*(uint32_t*)append((self), sizeof(uint32_t)) = (u))
+
+#include "modeveL-gen.h"
+
+// Hand-written functions {
+
+STATIC mp_obj_t _cmd0(mp_obj_t self, mp_obj_t n) {
+    C4(self, (0xffffff00 | mp_obj_get_int_truncated(n)));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(cmd0_obj, _cmd0);
+
+STATIC mp_obj_t _vertex2f(mp_obj_t self , mp_obj_t a0, mp_obj_t a1) {
+    int16_t x = (int16_t)(16 * mp_obj_get_float(a0));
+    int16_t y = (int16_t)(16 * mp_obj_get_float(a1));
+    C4(self, (0x40000000 | ((x & 32767) << 15) | (y & 32767)));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(vertex2f_obj, _vertex2f);
+
+STATIC mp_obj_t _cmd(size_t n_args, const mp_obj_t *args) {
```

Comment:
How is this different than `struct`? It'd be great to separate the implementation of these functions from the definition and arg parsing so that they are better documented. Right now, it is unclear what the `_eve` module would contain.

Reply:
* How is this different than `struct`?

The methods in this module (e.g. _vertex2f) align their arguments on bit boundaries, not the byte boundaries used by `struct`. Also these methods append the binary commands to an internal buffer, avoiding the overhead of constructing an ephemeral `bytes` object for the binary command.

 * It'd be great to separate the implementation of these functions from the definition and arg parsing so that they are better documented.

So taking `vertex2f`, which currently looks like this:

```
STATIC mp_obj_t _vertex2f(mp_obj_t self , mp_obj_t a0, mp_obj_t a1) {
    int16_t x = (int16_t)(16 * mp_obj_get_float(a0));
    int16_t y = (int16_t)(16 * mp_obj_get_float(a1));
    C4(self, (0x40000000 | ((x & 32767) << 15) | (y & 32767)));
    return mp_const_none;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_3(vertex2f_obj, _vertex2f);
```

Then I guess a split would be:

```
STATIC void vertex2f(mp_obj_t self, int16_x, int16_y) {
    C4(self, (0x40000000 | ((x & 32767) << 15) | (y & 32767)));
}

STATIC mp_obj_t _vertex2f(mp_obj_t self, mp_obj_t a0, mp_obj_t a1) {
    vertex2f(self,
             (int16_t)(16 * mp_obj_get_float(a0)),
             (int16_t)(16 * mp_obj_get_float(a1)));
    return mp_const_none;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_3(vertex2f_obj, _vertex2f);
```

Then `vertex2f` is in the business of packing integer bitfields for the hardware.
And `_vertex2f` is in the business of arg parsing before handing off to the integer interface.

(And similarly for all the other C methods.)

Is this along the right lines?

 * Right now, it is unclear what the _eve module would contain.

Modules `eve` and `_eve` together contain all the command definitions that are in the hardware datasheet. The split is that `_eve` contains the parts that benefit most from being coded in C.

Reply:
1. Ok, makes sense to use this for sub-byte packing. Note that `struct` can skip bytes objects with `pack_into` where you give it a bytearray.

2. Yup, that is on the right track. We then put the arg parsing in shared-bindings and the implementation in shared-modules. shared-bindings can also have inline docs. We name everything verbosely too so it includes the module name (`_eve` in this case). The example implementation function would be `common_hal__eve_vertex2f`. The arg parsing methods can be shorter because they are internal to the object.

3. Cool, sounds like the right structure then. It'd be great to document the C implementations for consistency with everything else.